### DISCUSSION TITLE
Fundamentos do Entity Framework > Operações Básicas > Required, MaxLenght e Column

### DISCUSSION THREAD

**Discussion by DrenokBR** [balta-io/forum#446] [Tira Dúvidas] [answered]
Fala André ! Bom Dia !

Ao invés de "[Required]", existe algum "Data Annotation" para definir uma coluna como "NULL", obrigatoriamente? E... sei que é uma prática comum... mas, não faria mais sentido usar um "Data Annotation" para definir uma coluna como "NULL" já que por "default" as colunas já são "NOT NULL" ?!?!

**Comment by andrebaltieri** [comment, marked as answer]
Boa tarde, @DrenokBR, como vai?

Caso não anote a propriedade como `Required`, o campo será **nulo**. Este é o processo padrão.
Pode realizar um teste e gerar as tabelas usando o `Migrations` para verificar.

**Reply by malighieri** [reply]
Acredito que ele pode estar se referindo a campos de tipos primitivos como int, decimal, ou mesmo algum campo cujo tipo seja um enumerado. Nestes casos, pelos campos não aceitarem null, eles serão não nulos por padrão. A menos que se use ? junto do tipo para indicar que ele pode ser nulo, 
No caso de campos string, dado que este tipo, por padrão, pode ser nulo, caso não se deseje isso, se faz necessário o uso da anotação Required

**Reply by andrebaltieri** [reply]
@malighieri bom dia!! Realmente, talvez seja isto!

Neste caso, como funciona o .NET (E outras linguagens também) 👇

Existem os tipos chamados de primitivos, que são tipos básicos (byte, int, float, string, char...)

Estes tipos também são conhecidos como tipos de valor (Value Types), por armazenarem um dado em si e não a referência para uma outra informação (Reference Types).

Dado que estes tipos tem seus respectivos tamanhos fixos (Veja aulas do curso de fundamentos do C# -- Value Types), eles sempre irão conter um valor.

Linguagens como o C (Em suas primeiras versões), não possuiam um valor padrão para estes tipos, então ocorria que durante a inicialização da variável, podia conter alguma "sujeira".

Isto ocorre por que quando declaramos `int i;`, alocamos 32 bits na memória para armazenar algum valor numérico inteiro. Porém, a memória é como uma planilha de Excel, que está em constante uso, toda preenchida.

Para otimizar a alocação, a memória nunca é apagada e sim sobrescrita (Isto acontece com o HD também -- Memória de longa duração), ou seja, quando pedimos 32 bits para alocar um inteiro, recebemos apenas o endereço da memória onde ele será alocado.

No caso do C, na declaração da variável, como um valor padrão não era atribuído, isto podia ocorrer:
```c
 int i;
printf("%s", i); // Irá imprimir o que estava na memória... valor aleatório (Inclusive podia dar erro)
```

Então tornou-se normal (Uma boa prática) inicializar as variáveis:
```c
 int i = 0;
printf("%s", i); // Agora tudo certo!
```

O C#, sendo uma linguagem de mais alto nível, já faz isto para nós, ou seja, todos os tipos primitivos tem um valor padrão definido, igual mostrei no curso de fundamentos do C#.

Desta forma, o mesmo código em C#, ficaria assim:
```csharp
int i;
Console.Write(i); // Vem zero por padrão
```

No caso dos Nullables, sempre que usamos `int?` ou qualquer outro tipo com `?`, significa que ele pode ser nulo.

Isto cria um objeto complexo do tipo `Nullable<T>` que possui uma propriedade extra chamada `HasValue`, determinando se a variável tem ou não algum valor assimilado.

Então o `int?` é apenas um syntax sugar para `Nullable<int>` que é um tipo complexo contendo um tipo inteiro e um booleano chamado `HasValue`.

Era esta a dúvida?

**Reply by malighieri** [reply]
Então..., não era uma dúvida. Eu tenho conhecimento destas características tanto de C quanto de C# com relação aos tipos primitivos e tipos de referência. Eu quis foi tentar mostrar porque @DrenokBR teve o questionamento de que, se o padrão para os campos é não ser nulo, faria sentido ter uma anotação para indicar que o campo poderia ser nulo. Entendi que ele se deparou com campos de tipos primitivos (int, decimal, etc...) que ao serem mapeados pelo EF no banco, geram sempre como não nulos, pelas características que você citou. Pra eles a anotação Required não faz sentido. E ficaria sem ter como indicar para estes campos que eles podem ser nulos. Ainda retomando sua explicação. Caso tenhamos campos de tipos primitivos que precisamos que aceitem nulo ao ser mapeados no banco, devemos indicar na sua declaração que eles podem ser nulos no C#, o que é feito com o ?. Sendo assim, não precisamos de uma anotação para isso. E por fim, mais uma vez, você explicou, para campos de tipos de referência, como string, o EF vai mapeá-lo por padrão como aceitando nulo, e para ele se faz necessário o uso da anotação Required quando se deseja que ele seja mapeado como não nulo.

**Reply by andrebaltieri** [reply]
Bom dia, @malighieri, desculpa a explicação, acreditei que fosse uma dúvida!
Muito obrigado pela ajuda e pelos complementos que colocou agora.

Valeu demais 🚀

**Reply by DrenokBR** [reply]
Boa tarde, Baltieri e Malighieri !

Poxa, vocês enriqueceram bastante minha pauta de estudos, pena que perdi esse "flow" (rsrs...). Gratidão.

Não lembro bem, mas acho que considerei que os campos eram "não nulos" por padrão por ter visto algo nesse sentido estudando a linguagem SQL. Pelo que Eu entendi, quando estamos escrevendo scripts SQL, parece que os campos são NOT NULL por padrão e, pelo que entendi, os comportamentos do SQL Server não possuem correspondência idêntica quando estamos usando o paradígma CODE FIRST. Seria isso mesmo ? Ou me confundi ?

**Reply by andrebaltieri** [reply]
@DrenokBR como comentei acima, o comportamento padrão é criar o campo como NULO, até por que se adicionar uma coluna a uma tabela existem com comportamento diferente disso, não funcionaria (Não sem o DEFAULT).
